Add documentation for Model Context Protocol Server integration (#36577)
home-assistant/home-assistant.github.io: source/_integrations/mcp_server.markdown

---
title: Model Context Protocol Server
description: Instructions on how to add a Model Context Protocol Server to Home Assistant.
ha_category:
  - Voice
ha_release: 2025.2
ha_iot_class: Local Push
ha_config_flow: true
ha_codeowners:
  - 'allenporter'
ha_domain: mcp_server
ha_integration_type: integration
related:
  - docs: /integrations/conversation/
    title: Conversation
---

The [Model Context Protocol](https://modelcontextprotocol.io) is an open protocol that standardizes how applications provide context to <abbr title="Large Language Models">LLMs</abbr>. The **Model Context Protocol Server** (MCP) integration enables using Home Assistant to provide context for <abbr title="Model Context Protocol">MCP</abbr> LLM Client Applications. For example, you can expose your Google Tasks To-do list as a tool for Claude Desktop.

Controlling Home Assistant is done by providing <abbr title="Model Context Protocol">MCP</abbr> clients access to the Assist API of Home Assistant. You can control what devices and entities it can access from the {% my voice_assistants title="exposed entities page" %}.

## Prerequisites

> 1. You need an [MCP client](https://modelcontextprotocol.io/clients) LLM Application such as [Claude for Desktop](https://claude.ai/download).
> 2. Since most clients do not support native remote servers, you need an additional local MCP server remote gateway.

Reviewer: ```suggestion
- You need an [MCP client](https://modelcontextprotocol.io/clients) LLM Application such as [Claude for Desktop](https://claude.ai/download).
- Since most clients do not support native remote servers, you need an additional local MCP server remote gateway.
```

> See [Client Configuration](#client-configuration) below for detailed instructions.

Reviewer: ```suggestion
For detailed configuration instructions, refer to the [Client configuration](#client-configuration) section.
```

[Server-Sent Events (SSE) transport](https://modelcontextprotocol.io/docs/concepts/transports#server-sent-events-sse)
allowing streaming client-to-server communication. Most MCP clients today only support
[stdio](https://modelcontextprotocol.io/docs/concepts/transports#standard-input-output-stdio) transport,
and directly run an MCP server as a local command line tool. You can 
use an MCP proxy server like [mcp-proxy](https://github.com/sparfenyuk/mcp-proxy)
to act as a gateway to the Home Assistant MCP SSE server.

## Client configuration

The Model Context Protocol specification does not yet define standards
for authentication and connecting to remote servers. These are a *work in progress*
and this configuration will likely change in the near future.

### Access control

For now, we can use
[Long-lived access token](https://developers.home-assistant.io/docs/auth_api/#long-lived-access-token) to control access to the API.

1. Visit your account profile settings, under the **Security** tab. {% my profile badge %}.

2. Create a **Long-lived access token**

3. Copy the access token to use when configuring the MCP client LLM application.

> See [Authentication](https://www.home-assistant.io/docs/authentication/#your-account-profile) for more information about Authentication in Home Assistant.

Reviewer: ```suggestion
For more information about Authentication in Home Assistant, refer to the [Authentication documentation](https://www.home-assistant.io/docs/authentication/#your-account-profile).
```

### Access control

For now, we can use
[Long-lived access token](https://developers.home-assistant.io/docs/auth_api/#long-lived-access-token) to control access to the API.

1. Visit your account profile settings, under the **Security** tab. {% my profile badge %}.

2. Create a **Long-lived access token**

3. Copy the access token to use when configuring the MCP client LLM application.

See [Authentication](https://www.home-assistant.io/docs/authentication/#your-account-profile) for more information about Authentication in Home Assistant.

### Example: Claude for Desktop

See [MCP Quickstart: For Claude Desktop Users](https://modelcontextprotocol.io/quickstart/user#for-claude-desktop-users)
for a detailed guide on using Claude for Desktop with an MCP server. It is recommended
to get the example server working first before using the Home Assistant MCP Server.

Claude for Desktop currently only supports local MCP servers using the [stdio](https://modelcontextprotocol.io/docs/concepts/transports#standard-input-output-stdio)
transport, run as a local command line tool. You can use a local MCP proxy server
to allow Claude for Desktop to access Home Assistant using the SSE transport.

> 1. Download [Claude for Desktop](https://claude.ai/download). 
> 
> 2. Install `mcp-proxy` following the instructions in the [README](https://github.com/sparfenyuk/mcp-proxy).
>    For example, `uv tool install git+https://github.com/sparfenyuk/mcp-proxy`.
> 
> 3. Open the configuration file. Visit **Settings…** and on the **Developer** tab, select **Edit Config**.
>    which will edit `claude_desktop_config.json`. The full file location depends on your
>    operating system (macOS or Windows).
> 
> 4. Add a new MCP server to the JSON file. You need to set the `SSE_URL` to the URL you use to
>    connect to Home Assistant with the path `/mcp_server/sse`. You will also need to set `API_ACCESS_TOKEN`
>    to the long live access token created above in the [access control instructions](#access-control)
> 
>   ```json
>   {
>     "mcpServers": {
>       "Home Assistant": {
>         "command": "mcp-proxy",
>         "env": {
>           "SSE_URL": "http://localhost:8123/mcp_server/sse",
>           "API_ACCESS_TOKEN": "<your_access_token_here>"
>         }
>       }
>     }
>   }
>   ```
> 
> 5. Restart Claude
> 
> 6. You will see a connection icon {% icon "mdi:connection" %} if things are set up correctly. Clicking the connection icon will show enabled MCP servers such as *Home Assistant*.
> 
> 7. You can then use Claude to control Home Assistant similar to how you control Home Assistant through the Voice Assistant. Claude wil ask you for permission before calling any tools.
> 
>   ![Screenshot of Claude for Desktop adding an item to a Home Assistant To-do list](/images/integrations/mcp_server/claude-todo-list-control.png)
> 

Reviewer: ```suggestion
1. Download [Claude for Desktop](https://claude.ai/download). 
2. Install `mcp-proxy` following the instructions in the [README](https://github.com/sparfenyuk/mcp-proxy).
   For example, `uv tool install git+https://github.com/sparfenyuk/mcp-proxy`.
3. Open the configuration file. Visit **Settings…** and on the **Developer** tab, select **Edit Config**.
   which will edit `claude_desktop_config.json`. The full file location depends on your
   operating system (macOS or Windows).
4. Add a new MCP server to the JSON file. You need to set the `SSE_URL` to the URL you use to
   connect to Home Assistant with the path `/mcp_server/sse`. You will also need to set `API_ACCESS_TOKEN`
   to the long live access token created above in the [access control instructions](#access-control)
    ```json
    {
      "mcpServers": {
        "Home Assistant": {
          "command": "mcp-proxy",
          "env": {
            "SSE_URL": "http://localhost:8123/mcp_server/sse",
            "API_ACCESS_TOKEN": "<your_access_token_here>"
          }
        }
      }
    }
    ```
5. Restart Claude.
6. You will see a connection icon {% icon "mdi:connection" %} if things are set up correctly. Clicking the connection icon will show enabled MCP servers such as *Home Assistant*.
7. You can then use Claude to control Home Assistant similar to how you control Home Assistant through the Voice Assistant. Claude wil ask you for permission before calling any tools.

  ![Screenshot of Claude for Desktop adding an item to a Home Assistant To-do list](/images/integrations/mcp_server/claude-todo-list-control.png)

```

---
title: Model Context Protocol Server
description: Instructions on how to add a Model Context Protocol Server to Home Assistant.
ha_category:
  - Voice
ha_release: 2025.2
ha_iot_class: Local Push
ha_config_flow: true
ha_codeowners:
  - 'allenporter'
ha_domain: mcp_server
ha_integration_type: integration
related:
  - docs: /integrations/conversation/
    title: Conversation
---

> The [Model Context Protocol](https://modelcontextprotocol.io) is an open protocol that standardizes how applications provide context to LLMs. The **Model Context Protocol Server** integration enables using Home Assistant to provide context for MCP LLM Client Applications. For example, you can expose your Google Tasks To-do list as a tool for Claude Desktop.

Reviewer: ```suggestion
The [Model Context Protocol](https://modelcontextprotocol.io) is an open protocol that standardizes how applications provide context to <abbr title="Large Language Models">LLMs</abbr>. The **Model Context Protocol Server** (MCP) integration enables using Home Assistant to provide context for <abbr title="Model Context Protocol">MCP</abbr> LLM Client Applications. For example, you can expose your Google Tasks To-do list as a tool for Claude Desktop.
```

> Controlling Home Assistant is done by providing MCP clients access to the Assist API of Home Assistant. You can control what devices and entities it can access from the {% my voice_assistants title="exposed entities page" %}.

Reviewer: ```suggestion
Controlling Home Assistant is done by providing <abbr title="Model Context Protocol">MCP</abbr> clients access to the Assist API of Home Assistant. You can control what devices and entities it can access from the {% my voice_assistants title="exposed entities page" %}.
```

> ## Pre-requisites

Reviewer: ```suggestion
## Prerequisites
```

> 1. You will need an [MCP client](https://modelcontextprotocol.io/clients) LLM Application such as [Claude for Desktop](https://claude.ai/download).

Reviewer: ```suggestion
1. You need an [MCP client](https://modelcontextprotocol.io/clients) LLM Application such as [Claude for Desktop](https://claude.ai/download).
```

> 1. Since most clients do not support native remote servers, you will need an additional local MCP server  remote gateway.

Reviewer: ```suggestion
2. Since most clients do not support native remote servers, you need an additional local MCP server remote gateway.
```

See [Client Configuration](#client-configuration) below for detailed instructions.

{% include integrations/config_flow.md %}

## Configuration options

The integration provides the following configuration options:

{% configuration_basic %}
Control Home Assistant:
  description: The API to use to expose tools over the Model Context Protocol. Clients
    can only control or provide information about entities that are [exposed](/voice_control/voice_remote_expose_devices/) to it.
{% endconfiguration_basic %}

> ## Architecture Overview

Reviewer: ```suggestion
## Architecture overview
```

client and prepares the available tools and passes them into the LLM with a prompt.

The Model Context Protocol follows a different pattern: An LLM application acts as
a client and can connect to multiple MCP servers to provide context. See the
[Model Context Protocol Introduction](https://modelcontextprotocol.io/introduction#general-architecture) for more details.

The Home Assistant Model Context Protocol Server integration implements the
[Server-Sent Events (SSE) transport](https://modelcontextprotocol.io/docs/concepts/transports#server-sent-events-sse)
allowing streaming client-to-server communication. Most MCP clients today only support
[stdio](https://modelcontextprotocol.io/docs/concepts/transports#standard-input-output-stdio) transport,
and directly run an MCP server as a local command line tool. You can 
use an MCP proxy server like [mcp-proxy](https://github.com/sparfenyuk/mcp-proxy)
to act as a gateway to the Home Assistant MCP SSE server.

> ## Client Configuration

Reviewer: ```suggestion
## Client configuration
```

> The Model Context Protocol specification does not yet currently define standards

Reviewer: ```suggestion
The Model Context Protocol specification does not yet define standards
```

for authentication and connecting to remote servers. These are a *work in progress*
and this configuration will likely change in the near future.

> ### Access Control

Reviewer: ```suggestion
### Access control
```

For now, we can use
[Long-lived access token](https://developers.home-assistant.io/docs/auth_api/#long-lived-access-token) to control access to the API.

1. Visit your account profile settings, under the **Security** tab. {% my profile badge %}.

> 1. Create a **Long-lived access token**

Reviewer: ```suggestion
2. Create a **Long-lived access token**
```
we use increasing-numbers for ordered lists in markdown

> 1. Copy the access token to use when configuring the MCP client LLM application.

Reviewer: ```suggestion
3. Copy the access token to use when configuring the MCP client LLM application.
```

See [Authentication](https://www.home-assistant.io/docs/authentication/#your-account-profile) for more information about Authentication in Home Assistant.

### Example: Claude for Desktop

See [MCP Quickstart: For Claude Desktop Users](https://modelcontextprotocol.io/quickstart/user#for-claude-desktop-users)
for a detailed guide on using Claude for Desktop with an MCP server. It is recommended
to get the example server working first before using the Home Assistant MCP Server.

Claude for Desktop currently only supports local MCP servers using the [stdio](https://modelcontextprotocol.io/docs/concepts/transports#standard-input-output-stdio)

> transport, run as a local command line tool. We can use a local MCP proxy server

Reviewer: ```suggestion
transport, run as a local command line tool. You can use a local MCP proxy server
```

to allow Claude for Desktop to access Home Assistant using the SSE transport.

1. Download [Claude for Desktop](https://claude.ai/download). 

> 1. Install `mcp-proxy` following the instructions in the [README](https://github.com/sparfenyuk/mcp-proxy).

Reviewer: ```suggestion
2. Install `mcp-proxy` following the instructions in the [README](https://github.com/sparfenyuk/mcp-proxy).
```

>    For example, `uv tool install git+https://github.com/sparfenyuk/mcp-proxy`

Reviewer: ```suggestion
   For example, `uv tool install git+https://github.com/sparfenyuk/mcp-proxy`.
```

> 1. Open the configuration file. Visit *Settings…* and *Developer* tab the click *Edit Config*

Reviewer: ```suggestion
3. Open the configuration file. Visit **Settings…** and on the **Developer** tab, select **Edit Config**.
```
We use bold to markup UI strings.

   which will edit `claude_desktop_config.json`. The full file location depends on your

>    operating system macOS or Windows.

Reviewer: ```suggestion
   operating system (macOS or Windows).
```

> 1. Add a new MCP server to JSON file. You will need to set the `SSE_URL` to the URL you use to

Reviewer: ```suggestion
4. Add a new MCP server to the JSON file. You need to set the `SSE_URL` to the URL you use to
```

   connect to Home Assistant with the path `/mcp_server/sse`. You will also need to set `API_ACCESS_TOKEN`
   to the long live access token created above in the [access control instructions](#access-control)

  ```json
  {
    "mcpServers": {
      "Home Assistant": {
        "command": "mcp-proxy",
        "env": {
          "SSE_URL": "http://localhost:8123/mcp_server/sse",
          "API_ACCESS_TOKEN": "<your_access_token_here>"
        }
      }
    }
  }
  ```

> 1. Restart Claude
> 
> 1. You will see a connection icon {% icon "mdi:connection" %} if things are set up correctly. Clicking the connection icon will show enabled MCP servers such as *Home Assistant*.
> 
> 1. You can then use Claude to control Home Assistant similar to how you control Home Assistant through the Voice Assistant. Claude wil ask you for permission before calling any tools.

Reviewer: ```suggestion
5. Restart Claude

6. You will see a connection icon {% icon "mdi:connection" %} if things are set up correctly. Clicking the connection icon will show enabled MCP servers such as *Home Assistant*.

7. You can then use Claude to control Home Assistant similar to how you control Home Assistant through the Voice Assistant. Claude wil ask you for permission before calling any tools.
```

#### Symptom: Failed to start MCP server: Could not start MCP server Home Assistant

When trying to configure a client like Claude for Desktop to talk to Home Assistant, the app shows a
message like "Failed to start MCP server: Could not start MCP server Home Assistant"

##### Description

This means that the local MCP server `mcp-proxy` could not start.

##### Resolution

Verify the command line arguments in the `claude_desktop_config.json` are correct. You may try to run
the command manually to verify that the command can be found.

#### Symptom: “MCP server Home Assistant disconnected” or "Could not attach to MCP server Home Assistant"

When trying to configure a client like Claude Desktop to talk to Home Assistant, the app shows a
message like "MCP server Home Assistant disconnected" or "Could not attach to MCP server Home Assistant".

##### Description

This means the MCP server has started, however the MCP server is having trouble communicating with Home Assistant,
or the MCP server in Home Assistant is not configured.

##### Resolution

To understand the root cause, first check debug logs on the client. For example in Claude for Desktop:

> 1. Visit *Settings...*
> 
> 1. Click *Developer*
> 
> 1. Click the `Home Assistant` MCP server
> 
> 1. Click *Open Logs Folder*
> 
> 1. View `mcp-server-Home Assistant.log`. These are known problems and their resolution:
> 
>    - `Client error '404 Not Found' for url 'http://localhost:8123/mcp_server/sse'`:
>      this means the MCP Server integration is not configured in Home Assistant.
> 
>    - `Client error '401 Unauthorized' for url 'http://localhost:8123/mcp_server/sse'`:
>      this means that the long live access token is not correct.

Reviewer: ```suggestion
1. Visit **Settings...**.
2. Select **Developer**.
3. Select the `Home Assistant` MCP server.
4. Select **Open Logs Folder**.
5. View `mcp-server-Home Assistant.log`. These are known problems and their resolution:
   - `Client error '404 Not Found' for url 'http://localhost:8123/mcp_server/sse'`:
     this means the MCP Server integration is not configured in Home Assistant.
   - `Client error '401 Unauthorized' for url 'http://localhost:8123/mcp_server/sse'`:
     this means that the long live access token is not correct.
```
